TailLight/pnp.cpp

```diff
@@ -0,0 +1,256 @@
+#include "driver.h"
+#include <Hidport.h>
+
+#include "pnp.h"
+
+// TODO: Streamline this once everything works
+typedef struct _SET_BLACK_WORK_ITEM_CONTEXT {
+    enum class state {
+        initialized = 0,
+        queued_waiting,
+        executing,
+        completed,
+        owning_device_unloaded
+    };
+
+    state _state;
+
+    WDFDEVICE owningDevice;
+    NTSTATUS  resultStatus;
+} SET_BLACK_WORK_ITEM_CONTEXT, * PSET_BLACK_WORK_ITEM_CONTEXT;
+
+EVT_WDF_WORKITEM                    EvtSetBlackWorkItem;
+EVT_WDF_REQUEST_COMPLETION_ROUTINE  EvtSetBlackCompletionRoutine;
+WDF_DECLARE_CONTEXT_TYPE_WITH_NAME(
+    SET_BLACK_WORK_ITEM_CONTEXT, 
+    Get_SetBlackWorkItemContext)
+
+PAGED_CODE_SEG
+NTSTATUS SetBlack(
+    WDFDEVICE device)
+/*++
+
+Routine Description:
+
+    Attempts to set the taillight to the default state of black.   
+
+Arguments:
+
+    Device - Handle to a framework device object.
+
+Return Value:
+
+    An NT or WDF status code.
+
+--*/
+{
+    TailLightReport* pReport = NULL;
+    NTSTATUS         status = STATUS_FAILED_DRIVER_ENTRY;
+    WDFREQUEST       request = NULL;
+    WDFIOTARGET      hidTarget = NULL;
+
+    WDFMEMORY InBuffer = 0;
+    BOOLEAN ret = FALSE;
+    BYTE* pInBuffer = nullptr;
+
+    PAGED_CODE();
+
+    KdPrint(("TailLight: %s\n", __func__));
+
+    // Set up a WDF memory object for the IOCTL request
+    WDF_OBJECT_ATTRIBUTES mem_attrib = {};
+    WDF_OBJECT_ATTRIBUTES_INIT(&mem_attrib);
+
+    WDF_REQUEST_SEND_OPTIONS sendOptions = {};
+    WDF_REQUEST_SEND_OPTIONS_INIT(
+        &sendOptions, 
+        WDF_REQUEST_SEND_OPTION_SYNCHRONOUS |
+        WDF_REQUEST_SEND_OPTION_TIMEOUT);
+    WDF_REQUEST_SEND_OPTIONS_SET_TIMEOUT(
+        &sendOptions, 
+        WDF_REL_TIMEOUT_IN_SEC(1));
+
+    hidTarget = WdfDeviceGetIoTarget(device);
+
+    //
+    // Create a report to send to the device.
+    //
+    pReport = (TailLightReport*)ExAllocatePool2(
```

Comment:
The `TailLightReport` struct is only 73bytes large so I suggest storing it on the stack so that you avoid the hassle of manual heap allocation.

Reply:
Thank you for addressing this.